src/csharp/Microsoft.Spark/Sql/Types/ComplexTypes.cs

```diff
@@ -69,9 +69,22 @@ private DataType FromJson(JObject json)
             return this;
         }
 
-        internal override bool NeedConversion() => true;
+        internal override bool NeedConversion() => ElementType.NeedConversion();
 
-        internal override object FromInternal(object obj) => throw new NotImplementedException();
+        internal override object FromInternal(object obj)
+        {
+            if (!NeedConversion())
+            {
+                return obj;
+            }
+
+            var convertedObj = new List<object>();
+            foreach(object o in (dynamic)obj)
```

Comment:
Why `dyanmic` ?

Reply:
`obj` could be `ArrayList`, `Date[]` , etc